File: src/modules/AlgoDatCourse/implemented/python/MaxHeap.py
class MaxHeap:

    def __init__(self, A):
        self.A = A
        self.heapSize = len(A)
        self.buildMaxHeap()

    """ HAUPTMETHODEN """

    def buildMaxHeap(self):

        self.heapSize = len(self.A)

        for i in range(self.heapSize // 2 - 1, -1, -1):
            self.maxHeapify(i)

    def maxHeapify(self, i):

        l = self.left(i)
        r = self.right(i)

        if l < self.heapSize and self.A[l] > self.A[i]:
            largest = l
        else:
            largest = i

        if r < self.heapSize and self.A[r] > self.A[largest]:
            largest = r

        if largest != i:

            temp = self.A[i]
            self.A[i] = self.A[largest]
            self.A[largest] = temp

            self.maxHeapify(largest)

    def maxHeapInsert(self, key):

        self.heapSize = len(self.A) + 1

        self.A.append(float("-inf"))

        self.maxHeapIncreaseKey(self.heapSize - 1, key)

    def extractMax(self):

        if self.heapSize == 0:
            raise Exception("Heap is empty")

        maximum = self.A[0]

        self.A[0] = self.A[self.heapSize - 1]

        self.heapSize -= 1

        self.A.pop()

        if self.heapSize > 0:
            self.maxHeapify(0)

        return maximum

    def maxHeapIncreaseKey(self, i, key):

        if key < self.A[i]:
            raise Exception(
                "Error: new key is smaller than current key"
            )

        self.A[i] = key

        while i > 0 and self.A[self.parent(i)] < self.A[i]:

            temp = self.A[i]
            self.A[i] = self.A[self.parent(i)]
            self.A[self.parent(i)] = temp

            i = self.parent(i)

    """ HILFSFUNKTIONEN IM BAUM """

    def left(self, i):
        return 2 * i + 1

    def right(self, i):
        return 2 * i + 2

    def parent(self, i):
        return (i - 1) // 2

    """ GETTER METHODEN """

    def getHeap(self):
        return self.A

    def getSortedArray(self):

        backupA = self.A.copy()
        backupHeapSize = self.heapSize

        sortedArray = [0] * self.heapSize

        for i in range(self.heapSize - 1, -1, -1):
            sortedArray[i] = self.extractMax()

        self.A = backupA
        self.heapSize = backupHeapSize

        return sortedArray

    def getMinimum(self):

        if self.heapSize == 0:
            raise Exception("Heap is empty")

        minimum = self.A[0]

        for i in range(self.heapSize // 2 - 1, self.heapSize):

            if self.A[i] < minimum:
                minimum = self.A[i]

        return minimum

    def getMaximum(self):

        if self.heapSize == 0:
            raise Exception("Heap is empty")

        return self.A[0]


if __name__ == "__main__":

    maxHeap = MaxHeap([5, 2, 1, 3, 7, 8])
    print(maxHeap.getHeap())
    print(maxHeap.getSortedArray())
    print(maxHeap.getMinimum())
    print(maxHeap.getMaximum())

File: src/modules/AlgoDatCourse/implemented/java/MaxHeap.java
/**
 * Eingabe:    Ein unsortiertes Array A von Zahlen, das noch keine Max-Heap-Eigenschaft hat.
 * Ausgabe:    Ein Array A, das als Max-Heap organisiert ist.
 * Methoden:   buildMaxHeap(), maxHeapify(i), maxHeapInsert(key), maxHeapIncreaseKey(i, key), extractMax()
 */
import java.util.Arrays;

public class MaxHeap {
    private int[] A;
    public int heapSize;

    MaxHeap(int[] A) {
        this.A = A;
        this.heapSize = A.length;
        buildMaxHeap();
    }

    /** HAUPTMETHODEN */

    public void buildMaxHeap() {
        heapSize = A.length;
        for (int i = heapSize / 2 - 1; i >= 0; i--) {
            maxHeapify(i);
        }
    }

    public void maxHeapify(int i) {
        int l = left(i);
        int r = right(i);
        int largest;

        if (l < heapSize && A[l] > A[i]) {
            largest = l;
        } else {
            largest = i;
        }

        if (r < heapSize && A[r] > A[largest]) {
            largest = r;
        }

        if (largest != i) {
            int temp = A[i];
            A[i] = A[largest];
            A[largest] = temp;
            maxHeapify(largest);
        }
    }

    public void maxHeapInsert(int key) {
        heapSize = A.length + 1;
        A = Arrays.copyOf(A, heapSize);
        A[heapSize - 1] = Integer.MIN_VALUE;
        maxHeapIncreaseKey(heapSize - 1, key);
    }

    public int extractMax() {
        if (heapSize == 0) {throw new IllegalStateException("Heap is empty"); }

        int max = A[0];
        A[0] = A[heapSize - 1];
        heapSize--;
        A = Arrays.copyOf(A, heapSize);
        if (heapSize > 0) {
            maxHeapify(0);
        }
        return max;
    }

    public void maxHeapIncreaseKey(int i, int key) {
        if (key < A[i]) {throw new IllegalArgumentException("Error: new key is smaller than current key"); }
       
        A[i] = key;
        while (i > 0 && A[parent(i)] < A[i]) {
            int temp = A[i];
            A[i] = A[parent(i)];
            A[parent(i)] = temp;
            i = parent(i);
        }
    }

    /** HILFSFUNKTIONEN IM BAUM */
    private int left(int i) {
        return 2 * i + 1;
    }

    private int right(int i) {
        return 2 * i + 2;
    }

    private int parent(int i) {
        return (i - 1) / 2;
    }

    /** GETTER METHODEN */
    public int[] getHeap() { return A; }
    public int[] getSortedArray() {
        int[] backupA = Arrays.copyOf(A, heapSize);
        int backupHeapSize = heapSize;
        int[] sortedArray = new int[heapSize];

        for (int i = heapSize - 1; i >= 0; i--) {
            sortedArray[i] = extractMax();
        }

        A = backupA;
        heapSize = backupHeapSize;
        return sortedArray;
    }

    public int getMinimum() {
        if (heapSize == 0) { throw new IllegalStateException("Heap is empty"); }

        int min = A[0];
        for (int i = heapSize/2-1; i < heapSize; i++) {
            if (A[i] < min) {
                min = A[i];
            }
        }
        return min;
    }
    public int getMaximum() {
        if (heapSize == 0) { throw new IllegalStateException("Heap is empty"); }
        return A[0];
    }
    

    public static void main(String[] args) {
        MaxHeap maxHeap = new MaxHeap(new int[]{5,2,1,3,7,8});
        System.out.println(Arrays.toString(maxHeap.getHeap()));
        System.out.println(Arrays.toString(maxHeap.getSortedArray()));
        System.out.println(maxHeap.getMinimum());
        System.out.println(maxHeap.getMaximum());
    }
}